From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dominic Frye <me@itsnebula.net>
Date: Wed, 21 May 2026 00:00:00 +0000
Subject: [PATCH] libcody: accept char8_t literals in S2C

When the host C++ compiler defaults to C++20 or later (e.g. GCC 16),
u8"" string literals have type const char8_t[] instead of const char[].
The existing S2C template only accepts char arrays, breaking the build
with errors like:

  error: no matching function for call to 'S2C(const char8_t [2])'

Add a char8_t overload guarded on __cpp_char8_t so older toolchains
still compile.
---
 libcody/cody.hh | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/libcody/cody.hh b/libcody/cody.hh
index 80370b5d925..a1f2dc04ee2 100644
--- a/libcody/cody.hh
+++ b/libcody/cody.hh
@@ -54,6 +54,15 @@ constexpr char S2C (char const (&s)[I])
   return s[0];
 }

+#if defined(__cpp_char8_t)
+template<unsigned I>
+constexpr char S2C (char8_t const (&s)[I])
+{
+  static_assert (I == 2, "only single octet strings may be converted");
+  return char (s[0]);
+}
+#endif
+
 /// Internal buffering class.  Used to concatenate outgoing messages
 /// and Lex incoming ones.
 class MessageBuffer
--
2.34.3
